### ISSUE TITLE
Feature Proposal: Reconsider & Refactor the Spatial Hash API

### ISSUE THREAD

**Issue by pushfoo** [pythonarcade/arcade#1605]
## Enhancement Proposal:
tl;dr reconsider the spatial hash API so people use arbitrary backing functions

This issue is intended to open the discussion to the wider community, including users not on Discord.

On Discord, einarf [proposed adding a GPU-backed SpatialHash object](https://discord.com/channels/458662222697070613/458662458198982676/1080470038177398784), which has since grown into a PR (#1604).

The initial proposal [set off a discussion](https://discord.com/channels/458662222697070613/458662458198982676/1080596415169966190) around adding support for other backing techniques for collision detection as well as fallbacks and user-configurable default providers (GPU, binary, then Python).

### What should be added/changed?

1. A Protocol type or ABC describing SpatialHash objects
2. Programming guide content covering when to use different providers and the general idea of writing your own

### What would the API look like?
Per [an earlier comment on a PR](https://github.com/pythonarcade/arcade/pull/1598#discussion_r1123782538), core methods include:
1. add
2. remove
3. reset
4. move

I'd also like to propose:
1. Move spatial hash recalculation methods onto the SpatialHash object
2. Consider renaming SpatialHash to something like CollisionSpace
3. Consider moving as well as some collision check methods onto it

### What would it help with?

It would allow users to implement their own backing SpatialHash techniques, such as quad trees or binding arbitrary binary-backed collision.

If we tweak SpriteLists a bit, we might also gain efficiency by  allowing advanced users to share a single SpatialHash between multiple lists. To my understanding, using an extra SpriteList to combine Sprites from other SpriteLists for the purpose of collision wastes time and space on re-creating GL resources in the extra SpriteList.

**Comment by einarf** [comment]
Actually my GPU collision is not even related to the spatial hash. I'd say we just rename the current one to `SpriteSpatialHash` or something and move on. It has already been refactored twice and we now have the simple `add`/`remove`/`move` api.

The current class works good enough as a base class for extending to be honest. We could make an ABC or a Protocol for it (in a perfect world).

1. I don't think this makes sense. The spatial hash either has a sprite or not. If you don't move the sprite it's a bug. However, it could have a method filling it with sprites from an interable.
2. It's a spatial hash and that's it. The name is already clear. It's used for other things than collision (like looking up nearby objects)
3. The spatial hash should not be responsible for anything other than what cell a sprite is in.

One thing that could be done is making a `arcade.spatial_hash` module had one for storing genric objects based on some Protocol, but this is nothing we will have time for in 3.0 unless someone wants to make a PR for that.

You can acually share spatial hashes between spritelists already, but this is somehing only and advacned user would care about.

> To my understanding, using an extra SpriteList to combine Sprites from other SpriteLists for the purpose of collision wastes time and space on re-creating GL resources in the extra SpriteList.

It does, but it's not that terrible. The spritelist can also be made lazy to avoid this. This issue is touchig so many issues I think there needs to be a lot more discussion before we dive into this can of worms or create any issues related to this. Maybe it's something we can solve in arcade 4.0, but right now it's very unclear.

TLDR; More discussion needed before we make issues out of this. For the simpler nice to have things it's just better to make a PR.
